Solved: 백준_피보나치 함수

2ssue opened the pull request:
# 백준 1003. 피보나치 함수 #32 

[문제 링크](https://www.acmicpc.net/problem/1003)

| 난이도 | 정답률(\_%) |
| :----: | :---------: |
| Silver 3 | 29.462% |

## 설계
- 계산해야할 자연수 N <= 40
- 피보나치 수열과 마찬가지로, 호출된 횟수도 `n = (n-1) + (n-2)` 형태이다. 따라서 배열에 각 수가 호출된 횟수를 넣어주면서 n까지 계산한다.
### 시간 복잡도
- O(N)
### 공간 복잡도
- 호출 횟수를 저장할 int형 배열 길이 41 
## 풀이
- 길이 41의 배열에 초기 값인 0, 1을 호출했을 때의 호출 횟수를 저장한다.
- 2부터 N까지 `n = (n-1) + (n-2)` 형태로 배열에 값을 저장해 return 한다. 
## 결과
| 메모리 | 실행 시간 |
| :----: | :---------: |
| 12988KB | 72ms |

___

함수로 따로 계산하도록 했는데 지금 보니 이전 케이스에서 그 값이 이미 계산되었을 수도 있네요. 그래도 회전이 큰 값이 아니라서 실행 시간에 큰 변화가 있을 것 같진 않습니다..
> 한번 해봤는데 더 느려졌네욤.. 하하😂 비교하는 작업이 매번 생겨서 그런 것 같습니다.

2ssue committed:
Solved: 백준_피보나치 함수

changicho commented on problems/baekjoon/1003/sujeong.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public static int[] getCallNumber(int n) {
+        int[][] callNumberOfZeroNOne = new int[41][2];
```
각 테스트 케이스별로 피보나치 dp를 매번 생성하는 방식이군요.

callNumberOfZeroNOne[][0], callNumberOfZeroNOne[][1] 로 각 수의 호출 회수를 구분하고 계신데요.

C에서의 구조체처럼 클래스를 선언해 배열로 선언하는건 어떨까요?

```java
public class Fibo {
  int zeros;
  int ones;
}

// ...

Fibo[] Fibos = new Fibo[41];

// 혹은 vector로

Vector<Fibo> vector = new Vector<Fibo>(41);
```

2ssue replied:
네 앞에서 계산된 값이 있으면 넘어가면 좋을 것 같은데, 계산이 오래걸리는 것도 아니라서 배보다 배꼽이 더 큰 것 같아 그냥 두었습니다..ㅎㅎ

클래스 선언은 좋은 방법인 것 같네요! 이 부분은 참고해서 한번 반영해보겠습니다 :)

2ssue committed:
Refactor: 이중 배열 형태에서 class 형태 배열로 변경

- 의미전달이 더 잘되는 형태인 class 방식으로 변경
- 단순히 구조체처럼 데이터만 저장하는 것보다는 다른 역할도 있는 것이
  자바가 지향하는 바에 더 맞을 것 같아서 String으로 변환하는 역할의
  함수를 하나 추가하여 String으로 변환하는 코드를 main에서 단순화하였음
